### PR TITLE
Fix nulls that are introduced during index partitioning

### PR DESCRIPTION
**Pull request by phofl**
cc @hendrikmakait 

Same as what we fixed in dask/dask before Christmas

ref #525

### CODE REVIEW

File: dask_expr/_shuffle.py [hunk 1]
```diff
@@ -244,7 +246,10 @@ def from_abstract_shuffle(cls, expr: Shuffle) -> Expr:
                 # Don't need to assign "_partitions" column
                 # if we are shuffling on a list of columns
                 nset = set(partitioning_index)
-                if nset & set(frame.columns) == nset and not index_shuffle:
+                if (
+                    nset & nset.intersection(set(frame.columns)) == nset
```
Comment (hendrikmakait) [inline dask_expr/_shuffle.py:250]: This condition feels overly complicated. If I'm not missing anything, this simplifies to `nset.issubset(set(frame.columns))`.

File: dask_expr/_merge.py [hunk 2]
```diff
@@ -284,12 +284,14 @@ def _lower(self):
                             left,
                             shuffle_left_on,
                             npartitions_out=left.npartitions,
+                            backend="tasks",
```
Comment (hendrikmakait) [inline dask_expr/_merge.py:287]: Why does this rely on `tasks`?

Comment (phofl) [reply dask_expr/_merge.py:287]: To keep the hashing consistent, not ideal but everything else loses data at the moment

Comment (phofl) [reply dask_expr/_merge.py]: Cleaned this up now after we removed the bug in dask/dask

File: dask_expr/_shuffle.py (hunk 1 above)
Comment (phofl) [reply dask_expr/_shuffle.py]: Removed altogether similar to dask/dask

**Comment by phofl** [comment]
Turns out this gave me so much trouble because dask/dask had a bug, I am disabling the release ci until the next release because we will get a bunch of failures there. Won't cut a new dask-expr release until dask/dask releases. We will have to bump the min dask version then as well

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -25,7 +25,7 @@ jobs:
       fail-fast: false
       matrix:
         python-version: ["3.9", "3.10", "3.11", "3.12"]
-        environment-file: [ci/environment.yml, ci/environment_released.yml]
```
Comment (hendrikmakait) [inline .github/workflows/test.yaml:28]: Is this intended?

**Review by hendrikmakait — APPROVED** [review]
Generally LGTM; one question that needs to be clarified before merging, but there's no need to go through another review. Thanks, @phofl!

File: .github/workflows/test.yaml (hunk 3 above)
Comment (phofl) [reply .github/workflows/test.yaml:28]: Yes

We are currently incompatible with released dask dask

Comment (hendrikmakait) [reply .github/workflows/test.yaml:28]: Makes sense, in an ideal world we should have split that into a separate PR.

Comment (phofl) [reply .github/workflows/test.yaml:28]: I actually merged this already in another PR before this one, not sure why the diff is still the way it is

**Pull request merged**